contracts/HybridCustody.cdc

```diff
@@ -0,0 +1,533 @@
+import "CapabilityFactory"
+import "CapabilityProxy"
+import "CapabilityFilter"
+
+/*
+HybridCustody defines a framework for sharing accounts via account linking. In the contract,
+you will find three main resources being used:
+1. ChildAccount - A resource which maintains an AuthAccount Capability, and handles publishing
+    and revoking access of that account via another resource called a ProxyAccount
+2. ProxyAccount - A second resource which exists on the same accounta as the ChildAccount. The ProxyAccount
+    resource is the capability which is shared to a parent account. Each proxy has its own set of rules
+    designating what Capability types can be retrieved, and what underlying type the capability points to can be
+    given.
+3. Manager - A resource setup by the parent which manages all child accounts shared with it. The Manager resource
+    also maintains a set of accounts that it "owns", meaning it has a capability to the full ChildAccount resource and
+    would then also be able to manage the child account's links as it sees fit.
+*/
+pub contract HybridCustody {
+
+    pub let ChildStoragePath: StoragePath
+    pub let ChildPublicPath: PublicPath
+    pub let ChildPrivatePath: PrivatePath
+
+    pub let ManagerStoragePath: StoragePath
+    pub let ManagerPublicPath: PublicPath
+    pub let ManagerPrivatePath: PrivatePath
+
+    pub let LinkedAccountPrivatePath: PrivatePath
+    pub let BorrowableAccountPrivatePath: PrivatePath
+
+    // TODO: Events!
+
+    // An interface which gets shared to a Manager when it is given full ownership of an account.
+    pub resource interface Account {
+        pub fun getAddress(): Address
+        pub fun isChildOf(_ addr: Address): Bool
+        pub fun getParentsAddresses(): [Address]
+        pub fun borrowAccount(): &AuthAccount?
+    }
+
+    // A ChildAccount shares the BorrowableAccount capability to itelf with ProxyAccount resources
+    pub resource interface BorrowableAccount {
+        access(contract) fun borrowAccount(): &AuthAccount
+    }
+
+    // Public methods anyone can call on a child account
+    pub resource interface ChildAccountPublic {
+        pub fun getParentsAddresses(): [Address]
+        pub fun getParentStatuses(): {Address: Bool}
+        pub fun getRedeemedStatus(addr: Address): Bool?
+
+        access(contract) fun setRedeemed(_ addr: Address)
+    }
+
+    // Accessible to the owner of the ChildAccount.
+    pub resource interface ChildAccountPrivate {
+        pub fun removeParent(parent: Address): Bool
+        pub fun publishToParent(
+            parentAddress: Address,
+            factory: Capability<&CapabilityFactory.Manager{CapabilityFactory.Getter}>,
+            filter: Capability<&{CapabilityFilter.Filter}>
+        )
+        pub fun giveOwnership(to: Address)
+        pub fun seal()
```

Comment:
Can you add a comment about what `seal()` means?

Reply:
Added a comment. We might want to remove this from the interface, though. It might only be relevant for this implementation of ChildAccount, thoughts?

Reply:
I agree, feels like adding it to the interface would be overly prescriptive. It's also only used in `giveOwnership()` in this implementation, so removing it from the interface wouldn't affect intended use here

Reply:
It's used by `giveOwnership` but also can be used on its own. This basically expresses two flavors of ownership:
1. The current owner transfers ownership of the account away to a new address
2. The current owner decides no one should own the account, and seals the account so that no one can add or remove any future resources, all they can do is interact with what's already there.

The second flavor of ownership is a much more trust-less approach where no one can alter the underlying paths or capabilities that are relied on. We want a way to let the owner exercise that function, but probably can find another way to expose it to the owner

Reply:
I see, that makes sense. Great idea. I don't think it's a problem to keep it in the interface